### ISSUE TITLE
Immediate cancellation of RPC calls leads to HTTP/2 RST packets for uninitialized streams

### ISSUE THREAD

**Issue by mfrischknecht** [grpc/grpc#19655]
### What version of gRPC and what language are you using?
From Nuget.org:
Grpc 1.22, Grpc.Core 1.22
 
### What operating system (Linux, Windows,...) and version?
Microsoft Windows 10
 
### What runtime / compiler are you using (e.g. python version or version of gcc)
.NET 4.7.2
Microsoft C# Tools 2.10.0 
(as distributed with VS2017 v. 15.9.8)

### What did you do?

We implemented a service that bundles and distributes long running client requests (using an internal protocol) to multiple upstream services. Said client requests occur in (sometimes rather large) batches of service method calls.

In the process of forwarding those calls, the intermediate service occasionally cancel newly created upstream calls without waiting for any replies (since some situation downstream or within the service renders the call unnecessary). 

In this case, we encountered `GOAWAY` signals from the service further upstream (implemented in Java), which seems to be the way Netty handles HTTP/2 `RST` signals for uninitialized streams.

I am able to consistently reproduce the issue using code akin to:

```c#
var service = new MyService.MyServiceClient(channel);
for (int i = 0; i < 20; ++i)
{
    Task.Run(async () => {
        using (var cancel = new CancellationTokenSource())
        using (var call = service.MyMethod(new MyRequest(), cancellationToken: cancel.Token))
        {
            var responses = call.ResponseStream;
            cancel.Cancel();
            while (await responses.MoveNext());
        }
    });
}
```
 
### What did you expect to see?

For any call that has not been communicated to the server yet, GRPC.Core should either initialize the according HTTP/2 stream before resetting it, or (probably more desirable for performance) any calls that are cancelled before being processed should not trigger any HTTP/2 communication.
 
### What did you see instead?
 
Using Wireshark, we were able to determine that GRPC.Core seems to send out HTTP/2 `RST` packages for streams that are not yet initialized (cf. packages 19 and 20), which causes the Netty-based upstream service to return a `GOAWAY` signal, rendering the entire connection unusable. [Section 6.4 of RFC 7540](https://tools.ietf.org/html/rfc7540#section-6.4) describes this as illicit behavior:

> RST_STREAM frames MUST NOT be sent for a stream in the "idle" state. If a RST_STREAM frame identifying an idle stream is received, the recipient MUST treat this as a connection error ([Section 5.4.1](https://tools.ietf.org/html/rfc7540#section-5.4.1)) of type PROTOCOL_ERROR.

(cf. also the [stream state diagram of section 5.1](https://tools.ietf.org/html/rfc7540#section-5.1)).

Capture (abbreviated:)

```
No.     Time           Source                Destination           Protocol Length Info
     16 0.119815       [client]              [server]              GRPC     198    HEADERS[5]: POST /[service]/[service_method], DATA[5] (GRPC) (PROTOBUF)
     17 0.120278       [server]              [client]              HTTP2    156    HEADERS[5]: 200 OK, DATA[5]
     18 0.120665       [client]              [server]              TCP      60     59650 → 22212 [ACK] Seq=778 Ack=305 Win=525056 Len=0
     19 0.121142       [client]              [server]              HTTP2    145    RST_STREAM[5], RST_STREAM[7], RST_STREAM[9], RST_STREAM[11], RST_STREAM[13], RST_STREAM[15], WINDOW_UPDATE[0]
     20 0.121368       [client]              [server]              HTTP2    67     RST_STREAM[17]
     21 0.121568       [server]              [client]              HTTP2    94     GOAWAY[0]
```

<details>
<summary>Complete capture</summary>
<pre><code>
No.     Time           Source                Destination           Protocol Length Info
      1 0.000000       [client]              [server]              TCP      66     59650 → 22212 [SYN, ECN, CWR] Seq=0 Win=8192 Len=0 MSS=1460 WS=256 SACK_PERM=1
      2 0.000023       [server]              [client]              TCP      66     22212 → 59650 [SYN, ACK] Seq=0 Ack=1 Win=29200 Len=0 MSS=1460 SACK_PERM=1 WS=128
      3 0.000459       [client]              [server]              TCP      60     59650 → 22212 [ACK] Seq=1 Ack=1 Win=525568 Len=0
      4 0.000878       [client]              [server]              HTTP2    153    Magic, SETTINGS[0], WINDOW_UPDATE[0], PING[0]
      5 0.000889       [server]              [client]              TCP      54     22212 → 59650 [ACK] Seq=1 Ack=100 Win=29312 Len=0
      6 0.001136       [client]              [server]              GRPC     392    HEADERS[1]: POST /[service]/Check, WINDOW_UPDATE[1], DATA[1] (GRPC) (PROTOBUF), WINDOW_UPDATE[0]
      7 0.001141       [server]              [client]              TCP      54     22212 → 59650 [ACK] Seq=1 Ack=438 Win=30336 Len=0
      8 0.001973       [server]              [client]              HTTP2    94     SETTINGS[0], WINDOW_UPDATE[0]
      9 0.002438       [server]              [client]              HTTP2    80     SETTINGS[0], PING[0]
     10 0.002707       [server]              [client]              HTTP2    173    HEADERS[1]: 200 OK, DATA[1]
     11 0.002791       [client]              [server]              TCP      60     59650 → 22212 [ACK] Seq=438 Ack=67 Win=525312 Len=0
     12 0.056455       [client]              [server]              TCP      60     59650 → 22212 [ACK] Seq=438 Ack=186 Win=525312 Len=0
     13 0.119339       [client]              [server]              GRPC     237    SETTINGS[0], HEADERS[3]: POST /[service]/[service_method], DATA[3] (GRPC) (PROTOBUF), WINDOW_UPDATE[0], PING[0]
     14 0.119551       [client]              [server]              HTTP2    67     RST_STREAM[3]
     15 0.119624       [server]              [client]              HTTP2    71     PING[0]
     16 0.119815       [client]              [server]              GRPC     198    HEADERS[5]: POST /[service]/[service_method], DATA[5] (GRPC) (PROTOBUF)
     17 0.120278       [server]              [client]              HTTP2    156    HEADERS[5]: 200 OK, DATA[5]
     18 0.120665       [client]              [server]              TCP      60     59650 → 22212 [ACK] Seq=778 Ack=305 Win=525056 Len=0
     19 0.121142       [client]              [server]              HTTP2    145    RST_STREAM[5], RST_STREAM[7], RST_STREAM[9], RST_STREAM[11], RST_STREAM[13], RST_STREAM[15], WINDOW_UPDATE[0]
     20 0.121368       [client]              [server]              HTTP2    67     RST_STREAM[17]
     21 0.121568       [server]              [client]              HTTP2    94     GOAWAY[0]
     22 0.121611       [server]              [client]              TCP      54     22212 → 59650 [FIN, ACK] Seq=345 Ack=882 Win=32512 Len=0
     23 0.121629       [client]              [server]              HTTP2    67     RST_STREAM[19]
     24 0.121809       [client]              [server]              HTTP2    67     RST_STREAM[21]
     25 0.122028       [server]              [client]              TCP      54     22212 → 59650 [RST, ACK] Seq=346 Ack=908 Win=32512 Len=0
     26 0.122058       [client]              [server]              TCP      60     59650 → 22212 [ACK] Seq=908 Ack=346 Win=525056 Len=0
     27 0.122075       [server]              [client]              TCP      54     22212 → 59650 [RST] Seq=346 Win=0 Len=0
     28 0.122184       [client]              [server]              HTTP2    67     RST_STREAM[1]
     29 0.122186       [server]              [client]              TCP      54     22212 → 59650 [RST] Seq=346 Win=0 Len=0
</code></pre>
</details>

The Netty service produces the following stack trace:

<details>
<summary>Netty stack trace</summary>
<pre><code>
[2019-07-16 10:02:03,218] [INFO ] [io.grpc.netty.NettyServerTransport.connections] [GrpcWorkerExecutor-9] Transport failed
io.netty.handler.codec.http2.Http2Exception: Stream 7 does not exist
        at io.netty.handler.codec.http2.Http2Exception.connectionError(Http2Exception.java:85)
        at io.netty.handler.codec.http2.DefaultHttp2ConnectionDecoder$FrameReadListener.verifyStreamMayHaveExisted(DefaultHttp2ConnectionDecoder.java:605)
        at io.netty.handler.codec.http2.DefaultHttp2ConnectionDecoder$FrameReadListener.onRstStreamRead(DefaultHttp2ConnectionDecoder.java:367)
        at io.netty.handler.codec.http2.Http2InboundFrameLogger$1.onRstStreamRead(Http2InboundFrameLogger.java:80)
        at io.netty.handler.codec.http2.DefaultHttp2FrameReader.readRstStreamFrame(DefaultHttp2FrameReader.java:516)
        at io.netty.handler.codec.http2.DefaultHttp2FrameReader.processPayloadState(DefaultHttp2FrameReader.java:260)
        at io.netty.handler.codec.http2.DefaultHttp2FrameReader.readFrame(DefaultHttp2FrameReader.java:160)
        at io.netty.handler.codec.http2.Http2InboundFrameLogger.readFrame(Http2InboundFrameLogger.java:41)
        at io.netty.handler.codec.http2.DefaultHttp2ConnectionDecoder.decodeFrame(DefaultHttp2ConnectionDecoder.java:148)
        at io.netty.handler.codec.http2.Http2ConnectionHandler$FrameDecoder.decode(Http2ConnectionHandler.java:378)
        at io.netty.handler.codec.http2.Http2ConnectionHandler.decode(Http2ConnectionHandler.java:438)
        at io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:502)
        at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:441)
        at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:278)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:374)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:360)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:352)
        at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1408)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:374)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:360)
        at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:930)
        at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:163)
        at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:682)
        at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:617)
        at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:534)
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:496)
        at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:906)
        at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
        at java.base/java.lang.Thread.run(Thread.java:834)
</code></pre>
</details>
 
### Anything else we should know about your project / environment?

Note: As this seems to be a timing problem and cancellations generally lead to exceptions within the C# gRPC API, I am best able to reproduce the issue **without using the Visual Studio debugger**. The debugger seems to slow things down considerably whenever any exceptions are in flight (I guess because it's signalling them all to the IDE and blocking the debugged process).

**Comment by mfrischknecht** [comment]
Also, in case anyone wonders about the `[square brackets]` within the capture above: I was ordered to redact system-internal information.

The `/[service]/Check` call is our internal health checking method, which I am leaving open in order to detect the `GOAWAY` signal within my test client (since all open calls are terminated when a `GOAWAY` signal is received).

**Comment by mfrischknecht** [comment]
Since the `stale` bot has marked this issue, I'd like to reiterate that this still is quite a big problem for our specific use case (although it likely is not all that typical for GRPC). We're using GRPC for continuously streaming live data, and the individual subscriptions can be triggered by user input, which means there's always the possibility of triggering this bug by accident. We also have to balance subscriptions between multiple connections with potentially small throughput (i.e. server capacity is not the problem), which can also lead to the same behavior.

It's also important to note that the delay between opening a request and cancelling it must not necessarily be that low in order to trigger this bug; as long as the request did not lead to an initialized HTTP/2 stream (which depends on the load of the library's internal task queue), this seems to happen rather deterministically. That means that if a user opens a couple of hundreds of subscriptions and cancels them again, they can still trigger this manually.

In either case, the entire connection to the server becoming unusable (along with any open subscriptions) is kind of a worst case scenario. So, for now, I had to avoid this issue by introducing a rather unfortunate hack: Our clients will not cancel any RPC call for which they have not received at least one response (since that is the only way I can be sure that the call has actually made it through the C implementation's internal queue).

**Comment by vjpai** [comment]
Although this triggers in the C# API, the underlying issue should be investigated as part of the C-core HTTP/2 transport. Reassigning to @yashykt - can you try to repro this behavior in a unit-test if it's deterministic as reported?

**Comment by firebook** [comment]
@yashykt any update?

**Comment by zhengdehui** [comment]
This bug is still existing in the latest C++ code of grpc-core, which is impacting all libraries basing on grpc-core, not only C#.

The cause is that: 
current design is using `grpc_chttp2_transport.outbuf` to send the payload of all streams in the transport. The logic of flushing data into `grpc_chttp2_transport.outbuf` is in `grpc/src/core/ext/transport/chttp2/transport/writing.cc`

~~~c++
grpc_chttp2_begin_write_result grpc_chttp2_begin_write(
    grpc_chttp2_transport* t) {
  WriteContext ctx(t);
  ctx.FlushSettings();
  ctx.FlushPingAcks();
  ctx.FlushQueuedBuffers();
  ctx.EnactHpackSettings();

  if (t->flow_control->remote_window() > 0) {
    ctx.UpdateStreamsNoLongerStalled();
  }

  /* for each grpc_chttp2_stream that's become writable, frame it's data
     (according to available window sizes) and add to the output buffer */
  while (grpc_chttp2_stream* s = ctx.NextStream()) {
    StreamWriteContext stream_ctx(&ctx, s);
    size_t orig_len = t->outbuf.length;
    stream_ctx.FlushInitialMetadata();
    stream_ctx.FlushWindowUpdates();
    stream_ctx.FlushData();
    stream_ctx.FlushTrailingMetadata();
    if (t->outbuf.length > orig_len) {
      /* Add this stream to the list of the contexts to be traced at TCP */
      s->byte_counter += t->outbuf.length - orig_len;
      if (s->traced && grpc_endpoint_can_track_err(t->ep)) {
        grpc_core::ContextList::Append(&t->cl, s);
      }
    }
    if (stream_ctx.stream_became_writable()) {
      if (!grpc_chttp2_list_add_writing_stream(t, s)) {
        /* already in writing list: drop ref */
        GRPC_CHTTP2_STREAM_UNREF(s, "chttp2_writing:already_writing");
      } else {
        /* ref will be dropped at end of write */
      }
    } else {
      GRPC_CHTTP2_STREAM_UNREF(s, "chttp2_writing:no_write");
    }
  }

  ctx.FlushWindowUpdates();

  maybe_initiate_ping(t);

  return ctx.Result();
}
~~~
When `grpc_chttp2_begin_write`, it `FlushQueuedBuffers` of grpc_chttp2_transport and then tries to iterate all data of streams which need to write data. However, the `ctx.NextStream()` cannot always return each stream because `grpc_chttp2_transport.outbuf` has a `target_write_size` limitation(1MB). see code below
~~~C++
  grpc_chttp2_stream* NextStream() {
    if (t_->outbuf.length > target_write_size(t_)) {
      result_.partial = true;
      return nullptr;
    }

    grpc_chttp2_stream* s;
    if (!grpc_chttp2_list_pop_writable_stream(t_, &s)) {
      return nullptr;
    }

    return s;
  }

/* How many bytes would we like to put on the wire during a single syscall */
static uint32_t target_write_size(grpc_chttp2_transport* /*t*/) {
  return 1024 * 1024;
}
~~~

This means in every invocation to `grpc_chttp2_begin_write`, it cannot make sure every stream is flushed in first stream iteration.

When some streams is not flushed in last round, the code will trigger next round by invoking `grpc_chttp2_begin_write` in `write_action_end_locked` so as to flush the left streams.
However, each time by invoking `grpc_chttp2_begin_write`, it will invoke `grpc_chttp2_begin_write` which will `ctx.FlushQueuedBuffers()` each time. If upper code call `grpc_chttp2_cancel_stream` in `chttp2_transport.cc`, it will append the `rst_stream` frame into `grpc_chttp2_transport.qbuf` which will be flushed to `grpc_chttp2_transport.outbuf` in the begining of next `grpc_chttp2_begin_write`. This could cause ths `rst_stream` of some streams to be flushed ahead of the `header` frame of that stream.

In summary, so when the load of requests is high, the output is limited by the `target_write_size` limitation(1MB) and has to seperate the streams into several round of `grpc_chttp2_begin_write`, and some `rst_stream` frame is sent shortly after the stream is created, this bug will happen.

From the code perspective, maybe this is caused by seperating the rst_stream from the StreamWriteContext object. If to handle rst_stream in the StreamWriteContext class, like `FlushInitialMetadata`, maybe this could be handle easily

**Comment by yashykt** [comment]
I think this is still a bug. @ctiller

**Comment by bytes-lost** [comment]
any update?

**Comment by jtattermusch** [comment]
I think realistically, the chances that this will get fixed in the future is very low:
- the issue has been open for quite a long time
- Grpc.Core has been using C-core from the LTS 1.46.x branch for the past year, so coming back to that code and fixing C-core issues has gotten more difficult.
- Grpc.Core is very close to getting deprecated (we're still evaluating the data, but the officially announced maintenance period is now over so even though we haven't made the deprecation official yet, we are likely to do that soon).

Since it's very unlikely that this will get fixed, I'll close this issue as "not actionable".

**Comment by segoon** [comment]
Hello,

As of Feb 2025, core is still not deprecated (new features are added to it) and still contains the bug. It is rather nasty in case of enabled deadline propagation and custom call cancellation. Can you probably look into the bug again, please?

**Comment by Anton3** [comment]
@yashykt @jtattermusch @ctiller notifying some of the members explicitly, otherwise notifications probably won't come from a closed issue

**Comment by yashykt** [comment]
@segoon / @Anton3 we are not adding new features to `Grpc.Core`. Are you perhaps referring to non-C# languages such as C++?

**Comment by Anton3** [comment]
Ah, got it. We've been noticing similar behavior with C++. We'll investigate a little bit further and create a new linked issue for C++ / grpc-core.
